pandas/io/sql.py

```diff
@@ -2008,6 +2023,188 @@ def _create_sql_schema(
+
+    def read_query(
+        self,
+        sql: str,
+        index_col: str | list[str] | None = None,
+        coerce_float: bool = True,
+        parse_dates=None,
+        params=None,
+        chunksize: int | None = None,
+        dtype: DtypeArg | None = None,
+        dtype_backend: DtypeBackend | Literal["numpy"] = "numpy",
+    ) -> DataFrame | Iterator[DataFrame]:
+        """
+        Read SQL query into a DataFrame. Keyword arguments are discarded.
+
+        Parameters
+        ----------
+        sql : str
+            SQL query to be executed.
+
+        Returns
+        -------
+        DataFrame
+
+        See Also
+        --------
+        read_sql_table : Read SQL database table into a DataFrame.
+        read_sql
+
+        """
+        with self.con.cursor() as cur:
+            return cur(sql).fetch_arrow_table().to_pandas()
+
+    read_sql = read_query
+
+    def to_sql(
+        self,
+        frame,
+        name: str,
+        if_exists: Literal["fail", "replace", "append"] = "fail",
+        index: bool = True,
+        index_label=None,
+        schema: str | None = None,
+        chunksize: int | None = None,
+        dtype: DtypeArg | None = None,
+        method: Literal["multi"] | Callable | None = None,
+        engine: str = "auto",
+        **engine_kwargs,
+    ) -> int | None:
+        """
+        Write records stored in a DataFrame to a SQL database.
+        Only frame, name, if_exists and schema are valid arguments.
+
+        Parameters
+        ----------
+        frame : DataFrame
+        name : string
+            Name of SQL table.
+        if_exists : {'fail', 'replace', 'append'}, default 'fail'
+            - fail: If table exists, do nothing.
+            - replace: If table exists, drop it, recreate it, and insert data.
+            - append: If table exists, insert data. Create if does not exist.
+        schema : string, default None
+            Name of SQL schema in database to write to (if database flavor
+            supports this). If specified, this overwrites the default
+            schema of the SQLDatabase object.
+        """
+        if schema:
+            table_name = f"{schema}.{name}"
+        else:
+            table_name = name
+
+        # TODO: pandas if_exists="append" will still create the
+        # table if it does not exist; ADBC has append/create
+        # as applicable modes, so the semantics get blurred across
+        # the libraries
+        mode = "create"
+        if self.has_table(name, schema):
+            if if_exists == "fail":
+                raise ValueError(f"Table '{table_name}' already exists.")
+            elif if_exists == "replace":
+                with self.con.cursor() as cur:
+                    cur.execute(f"DROP TABLE {table_name}")
+            elif if_exists == "append":
+                mode = "append"
+
+        import pyarrow as pa
+
+        tbl = pa.Table.from_pandas(frame)
```

Comment:
```suggestion
        tbl = pa.Table.from_pandas(frame, preserve_index=index)
```

That could honor the `index` keyword this way, I think

Otherwise could also do a reset_index, that might be easier to also implement the `index_label` keyword.

Reply:
Nice idea. For some reason though this produces a segfault with the latest development drivers. Here's a few lines from the traceback

```
#0  0x00007fff8a31aa5b in AdbcNsArrowSchemaDeepCopy (schema=0x140, schema_out=0x7fffffff8b00) at /home/willayd/clones/arrow-adbc/c/vendor/nanoarrow/nanoarrow.c:723
#1  0x00007fff8a307b28 in adbcpq::PostgresCopyStreamReader::GetSchema (this=0x0, out=0x7fffffff8b00)
    at /home/willayd/clones/arrow-adbc/c/driver/postgresql/postgres_copy_reader.h:962
#2  0x00007fff8a30187e in adbcpq::TupleReader::GetSchema (this=0x5555569c7b18, out=0x7fffffff8b00) at /home/willayd/clones/arrow-adbc/c/driver/postgresql/statement.cc:480
#3  0x00007fff8a302020 in adbcpq::TupleReader::GetSchemaTrampoline (self=0x5555569cfbf8, out=0x7fffffff8b00)
    at /home/willayd/clones/arrow-adbc/c/driver/postgresql/statement.cc:673
#4  0x00007fff93cfcf02 in arrow::(anonymous namespace)::ArrayStreamBatchReader::CacheSchema() const ()
   from /home/willayd/mambaforge/envs/adbc-testing/lib/python3.10/site-packages/pyarrow/../../../libarrow.so.1000
#5  0x00007fff93cfd2a6 in arrow::(anonymous namespace)::ArrayStreamBatchReader::ReadNext(std::shared_ptr<arrow::RecordBatch>*) ()
   from /home/willayd/mambaforge/envs/adbc-testing/lib/python3.10/site-packages/pyarrow/../../../libarrow.so.1000
#6  0x00007fff95376217 in __pyx_pw_7pyarrow_3lib_17RecordBatchReader_4read_next_batch(_object*, _object*) ()
   from /home/willayd/mambaforge/envs/adbc-testing/lib/python3.10/site-packages/pyarrow/lib.cpython-310-x86_64-linux-gnu.so
```

Reply:
On the sqlite side need to resolve https://github.com/apache/arrow-adbc/issues/905